python/cuml/fil/fil.pyx

```diff
@@ -0,0 +1,286 @@
+        self.num_nodes = self.calc_num_nodes()
+        for i in range(self.num_nodes):
+            node_info = <dense_node_t*>tree_node_info[i]
+            dense_node_init(<dense_node_t*> node_info,
+                            <float> weights[i],
+                            <float> self.threshold,
+                            <int> fid[i],
+                            <bool> def_left[i],
+                            <bool> is_leaf[i])
+        return self
+
+    def dense_node_decode(self, tree_node_info, weights,
+                          fid, def_left, is_leaf):
+
+        cdef uintptr_t weights_ptr, threshold_pointer, \
+            fid_pointer, def_left_pointer, is_leaf_pointer
+        cdef dense_node_t* node_info
+
+        for i in range(self.num_nodes):
+            weights_ptr = get_dev_array_ptr(weights[i])
+            threshold_pointer = get_dev_array_ptr(self.threshold)
+            fid_pointer = get_dev_array_ptr(fid[i])
+            def_left_pointer = get_dev_array_ptr(def_left[i])
+            is_leaf_pointer = get_dev_array_ptr(is_leaf[i])
+            node_info = <dense_node_t*>tree_node_info
+
+            dense_node_decode(<dense_node_t*> node_info,
+                              <float*> weights_ptr,
+                              <float*> threshold_pointer,
+                              <int*> fid_pointer,
+                              <bool*> def_left_pointer,
+                              <bool*> is_leaf_pointer)
+        return self
+
+    def init_dense(self, X):
+
+        cdef uintptr_t X_ptr
+        _, _, _, self.n_cols, self.dtype = \
+            input_to_dev_array(X, order='C')
+
+        cdef cumlHandle* handle_ =\
+            <cumlHandle*><size_t>self.handle.getHandle()
+
+        self.params.depth = self.depth
+        self.params.ntrees = self.num_trees
+        self.params.cols = self.n_cols
+        self.params.algo = self.algo
+        self.params.output = self.output
+        self.params.threshold = self.threshold
+
+        init_dense(handle_[0],
+                   <forest_t*> self.forest_pointer,
+                   <forest_params_t*> self.params)
+
+        return self
+
+    def predict(self, X):
+        cdef uintptr_t X_ptr
+
+        X_m, X_ptr, n_rows, _, _ = \
+            input_to_dev_array(X, order='C')
+
+        cdef cumlHandle* handle_ =\
+            <cumlHandle*><size_t>self.handle.getHandle()
+
+        preds = np.zeros(n_rows, dtype=np.int32)
+        cdef uintptr_t preds_ptr
+        preds_m, preds_ptr, _, _, _ = \
+            input_to_dev_array(preds)
+
+        predict(handle_[0],
+                self.forest_data,
+                <float*> preds_ptr,
+                <float*> X_ptr,
+                <size_t> n_rows)
+
+        return preds
+
+    def free(self):
+
+        cdef cumlHandle* handle_ =\
+            <cumlHandle*><size_t>self.handle.getHandle()
+
+        free(handle_[0],
+             self.forest_data)
+
+        return self
+
+
+class FIL(Base):
```

Comment:
This looks like just a wrapper to `FIL_impl`. Is it necessary?

Reply:
yes, since FILis statefull, the FIL class is the python API used by the user and the FIL_impl is the cython class with the implementation of the FIL functions.